src/Interpreters/maskSensitiveInfoInQueryForLogging.cpp

```diff
@@ -0,0 +1,607 @@
+                    ++arg_num; /// skip 'table' argument
+            }
+
+            /// Check if username and password are specified
+            /// (sharding_key can be of any type so while we're getting string literals they're username & password).
+            String username, password;
+            bool username_specified = tryGetStringFromArgument(function, arg_num, &username);
+            bool password_specified = username_specified && tryGetStringFromArgument(function, arg_num + 1, &password);
+
+            if (password_specified)
+            {
+                /// Password is specified so we do our replacement:
+                /// remote('addresses_expr', db.table, 'user', 'password', ...) -> remote('addresses_expr', db.table, 'user', '[HIDDEN]', ...)
+                wipePasswordFromArgument(function, data, arg_num + 1);
+            }
+        }
+
+        static void wipePasswordFromEncryptionFunctionArguments(ASTFunction & function, Data & data)
+        {
+            if constexpr (check_only)
+            {
+                data.can_contain_password = true;
+                return;
+            }
+
+            /// We replace all arguments after 'mode' with '[HIDDEN]':
+            /// encrypt('mode', 'plaintext', 'key' [, iv, aad]) -> encrypt('mode', '[HIDDEN]')
+
+            size_t num_arguments;
+            if (!tryGetNumArguments(function, &num_arguments) || (num_arguments < 2))
+                return;
+
+            wipePasswordFromArgument(function, data, 1);
+
+            auto & arguments = assert_cast<ASTExpressionList &>(*function.arguments).children;
+            arguments.resize(2);
+        }
+
+        static void visitBackupQuery(ASTBackupQuery & query, Data & data)
+        {
+            if (query.backup_name)
+            {
+                if (auto * backup_engine = query.backup_name->as<ASTFunction>())
+                    wipePasswordFromBackupEngineArguments(*backup_engine, data);
+            }
+
+            if (query.base_backup_name)
+            {
+                if (auto * base_backup_engine = query.base_backup_name->as<ASTFunction>())
+                    wipePasswordFromBackupEngineArguments(*base_backup_engine, data);
+            }
+        }
+
+        static void wipePasswordFromBackupEngineArguments(ASTFunction & engine, Data & data)
+        {
+            if (engine.name == "S3")
+            {
+                /// BACKUP ... TO S3(url, [aws_access_key_id, aws_secret_access_key])
+                wipePasswordFromArgument(engine, data, 2);
+            }
+        }
+
+        static void wipePasswordFromArgument(ASTFunction & function, Data & data, size_t arg_idx)
+        {
+            if constexpr (check_only)
+            {
+                data.can_contain_password = true;
+                return;
+            }
+
+            if (!function.arguments)
+                return;
+
+            auto * expr_list = function.arguments->as<ASTExpressionList>();
```

Comment:
can we use assert_cast here ?

Reply:
Then it will throw an exception if `function.arguments` is not `ASTExpressionList`. The function `maskSensitiveInfoInQueryForLogging()` is used to prepare a string for logging, so it's better not to throw exceptions in it (because otherwise we won't be able to log an invalid query). I'll add a comment